src/lib/storage/buffer/page_id.hpp

```diff
@@ -0,0 +1,62 @@
+#pragma once
+#include <magic_enum.hpp>
+#include "boost/integer/static_log2.hpp"
+
+namespace hyrise {
+
+#ifdef __APPLE__
+constexpr size_t OS_PAGE_SIZE = 16384;
+enum class PageSizeType { KiB16, KiB32, KiB64, KiB128, KiB256, KiB512, MiB1, MiB2 };
+#elif __linux__
+constexpr size_t OS_PAGE_SIZE = 4096;
+enum class PageSizeType { KiB4, KiB8, KiB16, KiB32, KiB64, KiB128, KiB256, KiB512, MiB1, MiB2 };
+#endif
+
+// Get the number of bytes for a given PageSizeType
+constexpr inline size_t bytes_for_size_type(const PageSizeType size) {
+  return OS_PAGE_SIZE << static_cast<size_t>(size);
+}
+
+constexpr size_t NUM_PAGE_SIZE_TYPES = magic_enum::enum_count<PageSizeType>();
+constexpr PageSizeType MIN_PAGE_SIZE_TYPE = magic_enum::enum_value<PageSizeType>(0);
+constexpr PageSizeType MAX_PAGE_SIZE_TYPE = magic_enum::enum_value<PageSizeType>(NUM_PAGE_SIZE_TYPES - 1);
+constexpr size_t PAGE_SIZE_TYPE_BITS = boost::static_log2<NUM_PAGE_SIZE_TYPES>::value + 1;
+
+/**
+ * PageIDs are used for addressing pages. They consist of a valid flag, a PageSizeType and an index. A Page ID can be unambiguously
+ * converted into a virtual address and back. If the valid flag is set, the page is stored in the buffer pool. Otherwise, the page
+ * does not exist or the virtual memory address is outside of the buffer pool.
+*/
+struct PageID {
+  using PageIDType = uint64_t;
+
+  PageIDType _valid : 1;
+  PageIDType _size_type : PAGE_SIZE_TYPE_BITS;
+  PageIDType index : (sizeof(PageIDType) * CHAR_BIT - PAGE_SIZE_TYPE_BITS - 1);
+
+  PageSizeType size_type() const {
+    return magic_enum::enum_value<PageSizeType>(this->_size_type);
+  }
+
+  size_t num_bytes() const {
+    return bytes_for_size_type(size_type());
+  }
+
+  bool valid() const {
+    return _valid;
+  }
+
+  PageID() = default;
+
+  constexpr PageID(const PageSizeType size_type, const PageIDType index, bool valid = true)
+      : _valid(valid), _size_type(static_cast<PageIDType>(size_type)), index(index) {}
```

Comment:
Different orders of parameters and members. Align them?

Reply:
Having the valid flag as the top bit, allows us to better read the binary representation. Still, we need it as the last parameter to set a default value.

Reply:
Ahh ... default parameter. Got it.